onnxruntime/contrib_ops/rocm/bert/multihead_attention.cu

```diff
@@ -26,21 +26,54 @@ namespace rocm {
       kRocmExecutionProvider,                                     \
       (*KernelDefBuilder::Create())                               \
           .TypeConstraint("T", DataTypeImpl::GetTensorType<T>()), \
-      MultiHeadAttention<T>);
-
-REGISTER_KERNEL_TYPED(float)
-REGISTER_KERNEL_TYPED(MLFloat16)
+      MultiHeadAttention<T>)
+
+REGISTER_MHA_KERNEL_TYPED(float);
+REGISTER_MHA_KERNEL_TYPED(MLFloat16);
+
+static constexpr int kPastSequenceLengthInputIndex = 7;
+static constexpr int kBeamWidthInputIndex = 8;
+static constexpr int kPastInputIndex = 5;
+static constexpr int kPresentOutputIndex = 1;
+
+#define REGISTER_DMMHA_KERNEL_TYPED(T)                                        \
+  ONNX_OPERATOR_TYPED_KERNEL_EX(                                              \
+      DecoderMaskedMultiHeadAttention,                                        \
+      kMSDomain,                                                              \
+      1,                                                                      \
+      T,                                                                      \
+      kRocmExecutionProvider,                                                 \
+      (*KernelDefBuilder::Create())                                           \
+          .MayInplace(kPastInputIndex, kPresentOutputIndex)                   \
+          .MayInplace(kPastInputIndex + 1, kPresentOutputIndex + 1)           \
+          .TypeConstraint("T", DataTypeImpl::GetTensorType<T>())              \
+          .InputMemoryType(OrtMemTypeCPUInput, kPastSequenceLengthInputIndex) \
+          .InputMemoryType(OrtMemTypeCPUInput, kBeamWidthInputIndex),         \
+      MultiHeadAttention<T>)
+
+REGISTER_DMMHA_KERNEL_TYPED(float);
+REGISTER_DMMHA_KERNEL_TYPED(MLFloat16);
 
 template <typename T>
 MultiHeadAttention<T>::MultiHeadAttention(const OpKernelInfo& info)
     : RocmKernel(info) {
+  if (info.node().OpType() == "MultiHeadAttention") {
+    attn_type_ = MHA;
+  } else if (info.node().OpType() == "DecoderMaskedMultiHeadAttention") {
+    attn_type_ = DMMHA;
+  } else {
+    ORT_THROW("unreachable");
+  }
+
   int64_t num_heads = 0;
   ORT_ENFORCE(info.GetAttr("num_heads", &num_heads).IsOK() && num_heads > 0);
   num_heads_ = static_cast<int>(num_heads);
 
   mask_filter_value_ = info.GetAttrOrDefault<float>("mask_filter_value", -10000.0f);
 
   scale_ = info.GetAttrOrDefault<float>("scale", 0.0f);
+
+  past_present_share_buffer_ = info.GetAttrOrDefault<int64_t>("past_present_share_buffer", 0LL);
```

Comment:
nit: past_present_share_buffer_ is bool and the attribute is int64. It is better to set like
past_present_share_buffer_ = (info.GetAttrOrDefault<int64_t>("past_present_share_buffer", 0LL) != 0LL).

Reply:
Done